File: apps/web/src/lib/api-client/analyst.ts
import { apiRequestJson, pathSegment, resolveApiBaseUrl } from "./client";
import type { TicketState } from "./tickets";
import type { TeamAvailability } from "./teams";

type AnalystAssignment = {
  id: string;
  analystUserId: string;
  assignedByUserId: string;
  route: "rfa" | "cm";
  createdAt: string;
  teamId: string;
  teamName: string | null;
};

export type WorkPackage = {
  id: string;
  title: string;
  status: "pending" | "complete";
  sortOrder: number;
};

export type AnalystTask = {
  ticketId: string;
  reference: string;
  state: TicketState;
  title: string;
  description: string | null;
  operationalQuestion: string | null;
  areaOrRegion: string | null;
  priority: string | null;
  requiredOutputFormat: string | null;
  chatSummary: string[];
  managerNotes: string[];
  assignments: AnalystAssignment[];
  workPackages: WorkPackage[];
  notes: { id: string; body: string; createdByUserId: string; createdAt: string }[];
  linkedProducts: {
    id: string;
    productId: string;
    reference: string;
    title: string;
    summary: string;
    createdAt: string;
  }[];
  drafts: {
    id: string;
    versionNumber: number;
    title: string;
    summary: string;
    productType: string;
    content: string;
    createdAt: string;
    assets: {
      id: string;
      name: string;
      assetType: string;
      mimeType: string;
      sizeBytes: number;
      sha256: string;
      detectedMimeType: string;
      previewKind: "pdf" | "image" | "text" | "metadata";
      processingStatus: string;
      previewAvailable: boolean;
    }[];
    description: string;
    sourceType: string;
    ownerTeam: string;
    areaOrRegion: string;
    classificationLevel: number;
    releasability: string[];
    handlingCaveats: string[];
    tags: string[];
    acgIds: string[];
    manifestHash: string;
  }[];
};

export type AnalystTaskList = {
  tasks: AnalystTask[];
};

export type AnalystConversation = {
  messages: {
    id: string;
    author: "user" | "assistant";
    body: string;
    createdAt: string;
  }[];
};

export type DraftProductInput = {
  title: string;
  summary: string;
  productType: string;
  content: string;
  assets: {
    name: string;
    assetType: string;
    mimeType: string;
    sizeBytes: number;
    sha256: string;
  }[];
};

export type ProductSubmissionMetadataInput = {
  title: string;
  summary: string;
  description: string;
  productType: string;
  sourceType: string;
  ownerTeam: string;
  areaOrRegion: string;
  classificationLevel: number;
  releasability: string[];
  handlingCaveats: string[];
  tags: string[];
  acgIds: string[];
  timePeriodStart: string | null;
  timePeriodEnd: string | null;
};

type AnalystCandidate = {
  userId: string;
  username: string;
  displayName: string;
};

export type AnalystCandidateList = {
  analysts: AnalystCandidate[];
};

export type AssignmentTeam = {
  teamId: string;
  name: string;
  kind: "rfa" | "cm";
};

export async function listAnalystTasks(): Promise<AnalystTaskList> {
  return apiRequestJson<AnalystTaskList>("/api/v1/analyst/tasks", { method: "GET" });
}

export async function getAnalystTaskConversation(ticketId: string): Promise<AnalystConversation> {
  return apiRequestJson<AnalystConversation>(
    `/api/v1/analyst/tasks/${pathSegment(ticketId)}/conversation`,
    { method: "GET" },
  );
}

export async function listAnalystCandidates(
  route: "rfa" | "cm",
  teamId: string,
): Promise<AnalystCandidateList> {
  const query = new URLSearchParams({ route, teamId });
  return apiRequestJson<AnalystCandidateList>(`/api/v1/analyst/candidates?${query.toString()}`, {
    method: "GET",
  });
}

export async function listAssignmentTeams(route: "rfa" | "cm"): Promise<AssignmentTeam[]> {
  const response = await apiRequestJson<{ teams: AssignmentTeam[] }>(
    `/api/v1/analyst/assignment-teams?route=${encodeURIComponent(route)}`,
    { method: "GET" },
  );
  return response.teams;
}

export function getAssignmentTeamAvailability(
  route: "rfa" | "cm",
  teamId: string,
  date: string,
): Promise<TeamAvailability> {
  const query = new URLSearchParams({ route, date });
  return apiRequestJson<TeamAvailability>(
    `/api/v1/analyst/assignment-teams/${pathSegment(teamId)}/availability?${query.toString()}`,
    { method: "GET" },
  );
}

export async function assignAnalystTask(
  ticketId: string,
  analystUserIds: string[],
  teamId: string,
  workPackages: string[],
  csrfToken: string,
): Promise<AnalystTask> {
  return apiRequestJson<AnalystTask>(`/api/v1/analyst/tasks/${pathSegment(ticketId)}/assign`, {
    body: JSON.stringify({ analystUserIds, teamId, workPackages }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function addAnalystNote(
  ticketId: string,
  body: string,
  csrfToken: string,
): Promise<AnalystTask> {
  return apiRequestJson<AnalystTask>(`/api/v1/analyst/tasks/${pathSegment(ticketId)}/notes`, {
    body: JSON.stringify({ body }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function linkAnalystProduct(
  ticketId: string,
  productId: string,
  csrfToken: string,
): Promise<AnalystTask> {
  return apiRequestJson<AnalystTask>(`/api/v1/analyst/tasks/${pathSegment(ticketId)}/products`, {
    body: JSON.stringify({ productId }),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function updateWorkPackage(
  ticketId: string,
  packageId: string,
  status: WorkPackage["status"],
  csrfToken: string,
): Promise<AnalystTask> {
  return apiRequestJson<AnalystTask>(
    `/api/v1/analyst/tasks/${pathSegment(ticketId)}/work-packages/${pathSegment(packageId)}`,
    {
      body: JSON.stringify({ status }),
      headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
      method: "PATCH",
    },
  );
}

export async function saveDraftProduct(
  ticketId: string,
  payload: DraftProductInput,
  csrfToken: string,
): Promise<AnalystTask> {
  return apiRequestJson<AnalystTask>(`/api/v1/analyst/tasks/${pathSegment(ticketId)}/drafts`, {
    body: JSON.stringify(payload),
    headers: { "Content-Type": "application/json", "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

export async function uploadAnalystProductSubmission(
  ticketId: string,
  metadata: ProductSubmissionMetadataInput,
  asset: File,
  csrfToken: string,
): Promise<AnalystTask> {
  const body = new FormData();
  body.set("metadata", JSON.stringify(metadata));
  body.set("asset", asset, asset.name);
  return apiRequestJson<AnalystTask>(
    `/api/v1/analyst/tasks/${pathSegment(ticketId)}/submissions/upload`,
    {
      body,
      headers: { "X-CSRF-Token": csrfToken },
      method: "POST",
    },
  );
}

export function workflowProductPreviewUrl(
  ticketId: string,
  versionId: string,
  assetId: string,
): string {
  return `${resolveApiBaseUrl()}/api/v1/workflow/products/${pathSegment(ticketId)}/versions/${pathSegment(versionId)}/assets/${pathSegment(assetId)}/preview`;
}

export async function submitTaskForReview(
  ticketId: string,
  csrfToken: string,
): Promise<AnalystTask> {
  return apiRequestJson<AnalystTask>(`/api/v1/analyst/tasks/${pathSegment(ticketId)}/submit`, {
    headers: { "X-CSRF-Token": csrfToken },
    method: "POST",
  });
}

File: apps/api/src/coeus/schemas/analyst.py
from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from coeus.domain.store import normalise_synthetic_release_markers
from coeus.schemas.tickets import ChatMessageResponse

WorkPackageText = Annotated[str, Field(min_length=3, max_length=180)]


class AnalystAssignmentRequest(BaseModel):
    analyst_user_ids: list[UUID] = Field(
        min_length=1,
        max_length=5,
        validation_alias="analystUserIds",
    )
    team_id: UUID = Field(validation_alias="teamId")
    work_packages: list[WorkPackageText] = Field(
        default_factory=list,
        max_length=8,
        validation_alias="workPackages",
    )


class AnalystNoteRequest(BaseModel):
    body: str = Field(min_length=3, max_length=2_000)


class AnalystConversationResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    messages: list[ChatMessageResponse]


class LinkProductRequest(BaseModel):
    product_id: UUID = Field(validation_alias="productId")


class WorkPackageUpdateRequest(BaseModel):
    status: str = Field(pattern="^(pending|complete)$")


class DraftAssetRequest(BaseModel):
    name: str = Field(min_length=3, max_length=180)
    asset_type: str = Field(min_length=2, max_length=80, validation_alias="assetType")
    mime_type: str = Field(min_length=3, max_length=120, validation_alias="mimeType")
    size_bytes: int = Field(gt=0, validation_alias="sizeBytes")
    sha256: str = Field(min_length=64, max_length=64)


class DraftProductRequest(BaseModel):
    title: str = Field(min_length=3, max_length=180)
    summary: str = Field(min_length=3, max_length=500)
    product_type: str = Field(min_length=3, max_length=80, validation_alias="productType")
    content: str = Field(min_length=10, max_length=20_000)
    assets: list[DraftAssetRequest] = Field(default_factory=list, max_length=5)


class ProductSubmissionMetadataRequest(BaseModel):
    title: str = Field(min_length=3, max_length=180)
    summary: str = Field(min_length=3, max_length=500)
    description: str = Field(min_length=3, max_length=2_000)
    product_type: str = Field(min_length=2, max_length=80, validation_alias="productType")
    source_type: str = Field(min_length=2, max_length=80, validation_alias="sourceType")
    owner_team: str = Field(min_length=2, max_length=80, validation_alias="ownerTeam")
    area_or_region: str = Field(min_length=2, max_length=180, validation_alias="areaOrRegion")
    classification_level: int = Field(ge=0, le=5, validation_alias="classificationLevel")
    releasability: list[str] = Field(min_length=1, max_length=12)
    handling_caveats: list[str] = Field(
        min_length=1, max_length=12, validation_alias="handlingCaveats"
    )
    tags: list[str] = Field(default_factory=list, max_length=30)
    acg_ids: list[UUID] = Field(min_length=1, max_length=20, validation_alias="acgIds")
    time_period_start: str | None = Field(default=None, validation_alias="timePeriodStart")
    time_period_end: str | None = Field(default=None, validation_alias="timePeriodEnd")

    def release_markers(self) -> tuple[tuple[str, ...], tuple[str, ...]]:
        return normalise_synthetic_release_markers(self.releasability, self.handling_caveats)


class AnalystCandidateResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    user_id: UUID = Field(serialization_alias="userId")
    username: str
    display_name: str = Field(serialization_alias="displayName")


class AnalystAssignmentResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    assignment_id: UUID = Field(serialization_alias="id")
    analyst_user_id: UUID = Field(serialization_alias="analystUserId")
    assigned_by_user_id: UUID = Field(serialization_alias="assignedByUserId")
    route: str
    created_at: datetime = Field(serialization_alias="createdAt")
    team_id: UUID | None = Field(serialization_alias="teamId")
    team_name: str | None = Field(serialization_alias="teamName")


class AssignmentTeamResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    team_id: UUID = Field(serialization_alias="teamId")
    name: str
    kind: str


class AssignmentTeamListResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    teams: list[AssignmentTeamResponse]


class WorkPackageResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    package_id: UUID = Field(serialization_alias="id")
    title: str
    status: str
    sort_order: int = Field(serialization_alias="sortOrder")


class AnalystNoteResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    note_id: UUID = Field(serialization_alias="id")
    body: str
    created_by_user_id: UUID = Field(serialization_alias="createdByUserId")
    created_at: datetime = Field(serialization_alias="createdAt")


class LinkedProductResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    link_id: UUID = Field(serialization_alias="id")
    product_id: UUID = Field(serialization_alias="productId")
    reference: str
    title: str
    summary: str
    created_at: datetime = Field(serialization_alias="createdAt")


class DraftAssetResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    asset_id: UUID = Field(serialization_alias="id")
    name: str
    asset_type: str = Field(serialization_alias="assetType")
    mime_type: str = Field(serialization_alias="mimeType")
    size_bytes: int = Field(serialization_alias="sizeBytes")
    sha256: str
    detected_mime_type: str = Field(serialization_alias="detectedMimeType")
    preview_kind: str = Field(serialization_alias="previewKind")
    processing_status: str = Field(serialization_alias="processingStatus")
    preview_available: bool = Field(serialization_alias="previewAvailable")


class DraftProductResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    version_id: UUID = Field(serialization_alias="id")
    version_number: int = Field(serialization_alias="versionNumber")
    title: str
    summary: str
    product_type: str = Field(serialization_alias="productType")
    content: str
    description: str
    source_type: str = Field(serialization_alias="sourceType")
    owner_team: str = Field(serialization_alias="ownerTeam")
    area_or_region: str = Field(serialization_alias="areaOrRegion")
    classification_level: int = Field(serialization_alias="classificationLevel")
    releasability: list[str]
    handling_caveats: list[str] = Field(serialization_alias="handlingCaveats")
    tags: list[str]
    acg_ids: list[UUID] = Field(serialization_alias="acgIds")
    manifest_hash: str = Field(serialization_alias="manifestHash")
    assets: list[DraftAssetResponse]
    created_at: datetime = Field(serialization_alias="createdAt")


class AnalystTaskResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    ticket_id: UUID = Field(serialization_alias="ticketId")
    reference: str
    state: str
    title: str
    description: str | None
    operational_question: str | None = Field(serialization_alias="operationalQuestion")
    area_or_region: str | None = Field(serialization_alias="areaOrRegion")
    priority: str | None
    required_output_format: str | None = Field(serialization_alias="requiredOutputFormat")
    chat_summary: list[str] = Field(serialization_alias="chatSummary")
    manager_notes: list[str] = Field(serialization_alias="managerNotes")
    assignments: list[AnalystAssignmentResponse]
    work_packages: list[WorkPackageResponse] = Field(serialization_alias="workPackages")
    notes: list[AnalystNoteResponse]
    linked_products: list[LinkedProductResponse] = Field(serialization_alias="linkedProducts")
    drafts: list[DraftProductResponse]


class AnalystTaskListResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    tasks: list[AnalystTaskResponse]


class AnalystCandidateListResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    analysts: list[AnalystCandidateResponse]
